### PR TITLE
[TheiaMeta] Binning with SemiBin2

### LINKED ISSUE #321: [Feature Request] Binning for TheiaMeta if reference is not provided
:cool:

### :pushpin: Explain the Request

As the next step for TheiaMeta development, binning can be implemented. 

### :books: Context

After an in-house assessment of binning performance, SemiBin2 was selected as the most suitable software for this purpose. It requires sorted bams of the metagenomic assembly files, and produces an array of fastas, one for each bin. 

### :chart_with_upwards_trend: Desired Behavior

If a reference is **not** provided, the metagenomic assembly goes through binning after mapping the read files to the assembly (with BWA). As output, an array of fasta files is provided to the Terra data table.

### PR DESCRIPTION
**Pull request by cimendes**
This PR closes #321

🗑️ This dev branch should be deleted after merging with main.

## :brain: Aim, Context and Functionality

Binning is the next logical step when it comes to metagenomic analysis through assembly and genomic characterization. It allows us to (ideally) separate the components of a community into their contigs.

Two processes are needed for binning:

1. The clean reads need to be mapped to the metagenomic assembly to produce abundance information for each contig 
2. The contigs and the sorted and indexed bam files are used by the binning algorithm to produce (ideally) as many bins as there are individual species in a community. 

Downstream characterization is not yet done in this PR.

## :hammer_and_wrench:  Impacted Workflows/Tasks & Changes Being Made
This will affect the behaviour of the workflow(s) even if users don’t change any workflow inputs relative to the last version : No

Running this workflow on different occasions could result in different results, e.g. due to the use of a live database, "latest" docker image, or stochastic data processing : Yes (binning is a stochastic algorithm and variations are expected)

#### 🔄 Data Processing 

An additional step has been introduced in the TheiaMeta workflow. Currently, this is a terminal step. After assembly, the resulting files are used to create a coverage report by mapping the clean reads to them. The resulting bams and the assembly file are binned with SemiBin2 to create possible multiple bin FASTA files.

A check was added in the SemiBin task to skip binning if the number of contigs over the minimum length threshold is less than two. This is to avoid failures with SemiBin software.

Docker/software or software versions changed: N/A

Databases or database versions changed: N/A

Data processing/commands changed: N/A

File processing changed: N/A

Compute resources changed: N/A

#### ➡️ Inputs 

New optional inputs:

- semibin cpu
- semibin disk_size
- semibin docker
- semibin evironment
- semibin mem  (@jrotieno should this be memory?)
- semibin min_len (@jrotieno should this be length?)
- semibin ratio

#### ⬅️ Outputs 

New outputs:

- semibin_bins (Array[File])
- semibin_version (String)
- semibin_docker (String

#### Test Dataset

Locally:

- 1 artificial metagenomic sample generated by concatenating in silico-generated reads of 15 target organisms

On Terra:

- 10 metagenomic samples from https://doi.org/10.1038/s41598-018-31873-w (Part of validation dataset)

#### Commandline Testing with MiniWDL or Cromwell (optional)

Semibin task was tested locally, concluding successfully

#### Terra Testing

image
https://app.terra.bio/#workspaces/theiagen-validations/Theiagen_Mendes_Sandbox/job_history/25c1eab1-4e9a-4390-b792-fc3e61daf519

#### Suggested Scenarios for Reviewer to Test

- Ensure the workflow functions using the reference track (should skip binning)

## :microscope: Final Developer Checklist

- [x] The workflow/task has been tested locally and results, including file contents, are as anticipated
- [x] The workflow/task has been tested on Terra and results, including file contents, are as anticipated
- [x] The CI/CD has been adjusted and tests are passing (to be completed by Theiagen developer)
- [x] Code changes follow the [style guide](https://theiagen.notion.site/Style-Guide-WDL-Workflow-Development-bb456f34322d4f4db699d4029050481c)

## 🎯 Reviewer Checklist 

- [ ] All impacted workflows/tasks have been tested on Terra with a different dataset than used for development
- [x] All reviewer-suggested scenarios have been tested and any additional
- [x] All changed results have been confirmed to be accurate
- [x] All workflows/tasks impacted by change/s have been tested using a standard validation dataset to ensure no unintended change of functionality
- [x] All code adheres to the style guide
- [x] MD5 sums have been updated
- [x] The PR author has addressed all comments

## 🗂️ Associated Documentation (to be completed by Theiagen developer)

- [x] Relevant documentation on the Public Health Resources "PHB Main" has been updated
- [x] Workflow diagrams have been updated to reflect changes

### CODE REVIEW

**Comment by jrotieno** [comment]
@cimendes yes, if you can make the `mem` memory and `len` length, that'd be great.

**Comment by cimendes** [comment]
> @cimendes yes, if you can make the `mem` memory and `len` length, that'd be great.

Done! Thank you!

**Comment by cimendes** [comment]
TODO:  merge main in!

File: tasks/assembly/task_semibin.wdl
Comment (kapsakcj) [inline tasks/assembly/task_semibin.wdl:48]: Question - are any of the other Semibin outputs useful to keep/save? Just wondering if there's any important output Strings to capture or Files to save as outputs

**Review by kapsakcj — APPROVED** [review]
Tested on Terra with 4 samples from validation dataset (I don't know where to get good test data, so I tested with what I have easy access to)

https://app.terra.bio/#workspaces/theiagen-validations/curtis-sandbox-theiagen-validations/job_history/bddc023a-2045-4a97-93f2-5bf058ff77be

Workflows ran as expected, and produced the semibin outputs without issue.

I have no further suggestions.

Nice work!

**Comment by kapsakcj** [comment]
Since this is a new tool added to TheiaMeta - will you add this to the workflow diagram and documentation? I think that may be the last thing needed before merging the PR. Please let me know when you're ready to merge and I can hit the button

**Comment by cimendes** [comment]
> Since this is a new tool added to TheiaMeta - will you add this to the workflow diagram and documentation? I think that may be the last thing needed before merging the PR. Please let me know when you're ready to merge and I can hit the button

yes!! I was waiting on a semi-approval to get that going :) Will update now

**Comment by cimendes** [comment]
@kapsakcj docs have been updated!

**Pull request merged**
